### ISSUE TITLE
mqtt messages

### ISSUE THREAD

**Issue by Sborkahome** [jakowenko/double-take#52]
David, hello!
Thank you for your project and for answering our questions quickly. But you can add a few more things:
1) messages error output in mqtt, for example, in config.yml the parameters are incorrect, or detectors are not specified;
2) you can made button "reset" in web-ui double-take.
Now all this can be done only through docker/portainer which is very inconvenient.

**Comment by jakowenko** [comment]
Hey @Sborkahome, thanks for your suggestions, these are great!

1. Would you like any error messages published to a MQTT topic? Did you have any preferences on the name or format or just something simple like `double-take/errors`?
2. Are you hoping to just restart the API form the UI? There currently is a localhost:PORT/#/config route, when you save changes here the API does restart, but I can move that functionality to other pages too.

**Comment by Sborkahome** [comment]
1. May be create double-take/errors = "description of the error", for example, <<syntax error, line 10>> or <<detector not found>>.
2. Yes, button "reset" in ui this cool!

**Comment by jakowenko** [comment]
Hey @Sborkahome, finally got around to implementing this.

Any errors from the API will be published to `double-take/errors`. Here's a sample of what one may look like.

Let me know if this will work.

```json
{
  "stack": "Error: compreface process error: Required part file is missing\n    at Object.module.exports.normalize (/double-take/api/src/util/detectors/compreface.js:71:11)\n    at module.exports.normalize (/double-take/api/src/util/detectors/actions/index.js:7:74)\n    at Object.module.exports.process (/double-take/api/src/util/process.util.js:93:33)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async Promise.all (index 0)\n    at async Object.module.exports.polling (/double-take/api/src/util/process.util.js:37:23)\n    at async Promise.all (index 0)\n    at async module.exports.start (/double-take/api/src/controllers/recognize.controller.js:139:70)",
  "message": "compreface process error: Required part file is missing"
}
```

**Comment by ozett** [comment]
this topic about mqtt is general FR,

so i would like to get the matched image with mqtt, like frigate itself does on snapshots.

may you/i make it a FR ?

![image](https://user-images.githubusercontent.com/5937811/132136981-833be2dd-36db-42fe-a6b6-8bd31b395635.png)

**Comment by jakowenko** [comment]
Hey @ozett,

I pushed a new beta build with support for this feature. A base64 encoded string will be returned from the API and published to the MQTT topics. Let me know if this will solve your issue 🍻 

The available options are `true`, `false`, or `box`.

`true` - source image
`false` - disabled
`box` - source image with bounding boxes of results

```yaml
save:
  base64: 'box'
```

**Comment by ozett** [comment]
thanks for putting this in that fast. Great! And Great work at all. 💐 

i pull the only "beta" called beta and now have this config:
```
save:
  matches: true
  unknown: true
  base64: 'box'

```
is there a subtopic for the image?  like ``` double-take/matches/image``` ?
i have this configuration:
```
mqtt:
  host: 192.168.14.204
  topics:
    matches: double-take/matches
    frigate: /frigate5/events

```

**Comment by ozett** [comment]
got one:
![image](https://user-images.githubusercontent.com/5937811/132321662-596ffa65-a9b7-4b16-b8fd-9253541b3592.png)

**Comment by jakowenko** [comment]
Awesome! Does it seem to be working for you? I figured it would be easier to have the base64 image on the same message rather than a separate one.

**Comment by ozett** [comment]
what a pity, i worked only this one time i showed above.
but the gui showed more matches. nothing came over mqtt.
i was to restart double-take, but may that is a bug?

i subscribed with mosquito on cli and got again the first and only message
 ```
 mosquitto_sub -v -t /double-take/#
/double-take/matches/misterX {"id":"1631006825.831465-jn1rlo","duration":10.31,"timestamp":"2021-09-07T09:27:16.580Z","
...
```

![image](https://user-images.githubusercontent.com/5937811/132395595-85a1a554-cc66-48ec-a271-3423c4605367.png)
what about the red-matches. should they also come over mqtt as they apear on the gui?
even the green one was not send.

**Comment by jakowenko** [comment]
Is that command right? That didn't work for me, but this showed MQTT messages coming in.

`mosquitto_sub -v -t 'double-take/#'`

**Comment by ozett** [comment]
my mqtt-topics are set to start with "/".... you have to adjust yours...
![image](https://user-images.githubusercontent.com/5937811/132450973-f1c7015b-301a-47c9-83cb-40445fd1e653.png)
[Littel Info how to subscribe to all topics](https://stackoverflow.com/questions/32904012/how-do-i-subscribe-to-all-topics-of-a-mqtt-broker)

**Comment by ozett** [comment]
there is a debug filter for mqtt-messages from double-take.
not all matches are coming in with mqtt. 
seems that some matches are missing. 
How do i debug this more? may you check the code?

![image](https://user-images.githubusercontent.com/5937811/132452044-0f7d9fbe-6f15-45ef-9f3f-5db5e1bef275.png)

FR: absolute time/date written right behind the past time since match maybe helpful

**Comment by ozett** [comment]
maybe mqtt can send every time? even "false-"matches?
i could than filter at the receiver in node-red if i want them..

![image](https://user-images.githubusercontent.com/5937811/132495030-3ca9cbf0-ed52-482a-b2f1-9f6847adbc33.png)


this are the flow-nodes:
![image](https://user-images.githubusercontent.com/5937811/132494271-a2a26b7e-7afa-4b7b-979e-06ec4f9e39d9.png)

```
[{"id":"ad8ddf2b.79d33","type":"mqtt in","z":"2d5205a0.26601a","name":"","topic":"/double-take/#","qos":"2","datatype":"auto","broker":"9b50c12.74a3a4","x":710,"y":900,"wires":[["87f6d4aa.1261e8"]]},{"id":"f7ab8316.5c808","type":"debug","z":"2d5205a0.26601a","name":"dt-mqtt-matches","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":1040,"y":900,"wires":[]},{"id":"87f6d4aa.1261e8","type":"json","z":"2d5205a0.26601a","name":"","property":"payload","action":"","pretty":false,"x":870,"y":900,"wires":[["f7ab8316.5c808","3fe2df70.4cf46"]]},{"id":"3fe2df70.4cf46","type":"image","z":"2d5205a0.26601a","name":"","width":"320","data":"payload.match.base64","dataType":"msg","thumbnail":false,"active":true,"pass":false,"outputs":0,"x":800,"y":960,"wires":[]},{"id":"9b50c12.74a3a4","type":"mqtt-broker","name":"192.168.14.204-hik","broker":"192.168.14.204","port":"1883","clientid":"","usetls":false,"compatmode":true,"keepalive":"60","cleansession":true,"birthTopic":"","birthQos":"0","birthPayload":"","closeTopic":"","closePayload":"","willTopic":"","willQos":"0","willPayload":""}]
```

💚   sending image with mqtt in node red:
```
http://192.168.14.14:3000/api/recognize?url=https://jakowenko.com/img/david.92f395c6.jpg
```
![image](https://user-images.githubusercontent.com/5937811/132494606-c6c8ee63-6d2e-436b-98ff-4f4628445e0a.png)
![image](https://user-images.githubusercontent.com/5937811/132494716-9a76faa9-4684-4d70-8ccf-991af4a2e757.png)




🟥   no image with mqtt in node-red . no data at all
( but at the same time it shows perfect in browser on the match-tab as a failed-match!!)

```
http://192.168.14.14:3000/api/recognize?url=https://avatars.githubusercontent.com/u/569905?v=4
```
![image](https://user-images.githubusercontent.com/5937811/132493256-4e980d5b-ce7b-492f-844d-02ffa56e57c1.png)
![image](https://user-images.githubusercontent.com/5937811/132493391-b9edeecf-853d-48b3-b24c-6a5ab926bf00.png)

**Comment by jakowenko** [comment]
Thanks for the detailed info, MQTT should be sent messages for every match/miss. I haven't experienced the inconsistencies you are talking about, but I will look more at that code tonight and try to get my setup similar to yours to compare. I'll report back what I find.

**Comment by ozett** [comment]
after some fiddling i have a working flow 
to get matching images via mqtt to node-red -> 
and to send me telegramm-faces

![image](https://user-images.githubusercontent.com/5937811/132540872-e1660a3d-6f7b-43a7-af5b-7dbecb4cca52.png)

**Comment by ozett** [comment]
> Thanks for the detailed info, MQTT should be sent messages for every match/miss.

at the moment i trigger on base64-image in the mqtt-flow. (assuming that there is always an image )
maybe for misses thats missing. i did not check this detail yet.

**Comment by Haloooch** [comment]
@ozett Don't mean to hijack this conversating and I know this is bug reporting not a forum, but any chance you can share the flow above. I am getting base64 images in all messages but when saving to disk the images are corrupt.

**Comment by ozett** [comment]
> @ozett Don't mean to hijack this conversating and I know this is bug reporting not a forum, but any chance you can share the flow above. I am getting base64 images in all messages but when saving to disk the images are corrupt.

me, too. took me some hours to get a working flow. i prepare shraring here. stay tuned

**Comment by ozett** [comment]
took me some time and kindly little help from telegram-node maintainer until the flow did not save the image, but sends from buffer. see that in the function-node.

![image](https://user-images.githubusercontent.com/5937811/132810404-b6c7f464-d2ac-4707-932a-3223d3cff8a3.png)


<details><summary>CLICK ME</summary>
<p>

```
[{"id":"7ebfb70b.c4dab8","type":"mqtt in","z":"3701c362.4ce7dc","name":"","topic":"/double-take/#","qos":"2","datatype":"auto","broker":"9b50c12.74a3a4","x":190,"y":780,"wires":[["d64b9790.47c938"]]},{"id":"d64b9790.47c938","type":"json","z":"3701c362.4ce7dc","name":"","property":"payload","action":"","pretty":false,"x":350,"y":780,"wires":[["51776b0e.1d0494"]]},{"id":"4b3423f9.64fa4c","type":"telegram sender","z":"3701c362.4ce7dc","name":"send","bot":"","haserroroutput":false,"outputs":1,"x":1270,"y":780,"wires":[["65cbc159.f2379"]]},{"id":"65cbc159.f2379","type":"debug","z":"3701c362.4ce7dc","name":"tlgr.send.photo","active":false,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":1440,"y":780,"wires":[]},{"id":"6f6f00ec.a7e07","type":"base64","z":"3701c362.4ce7dc","name":"","action":"","property":"payload","x":860,"y":780,"wires":[["f41350b2.a332a"]]},{"id":"7d1083a6.6f75cc","type":"change","z":"3701c362.4ce7dc","name":"img2payload","rules":[{"t":"move","p":"payload.match.base64","pt":"msg","to":"payload","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":710,"y":780,"wires":[["6f6f00ec.a7e07"]]},{"id":"51776b0e.1d0494","type":"change","z":"3701c362.4ce7dc","name":"save name,conf","rules":[{"t":"move","p":"payload.match.name","pt":"msg","to":"name","tot":"msg"},{"t":"move","p":"payload.match.confidence","pt":"msg","to":"confidence","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":520,"y":780,"wires":[["7d1083a6.6f75cc"]]},{"id":"f41350b2.a332a","type":"function","z":"3701c362.4ce7dc","name":"send picture as buffer","func":"var pl = {\n  content: msg.payload,\n  message: `Message has the AI picture from face-rec`,\n  type : 'photo',\n  chatId: \"-431012345\",\n  caption1: \"caption\",\n  caption: \"AI gesichtet: \"  + msg.name + \" (\" + msg.confidence + \"% )\",\n  options: {\"disable_notification\": true}\n}\n\nmsg.payload = pl;\n\n\n/* type can be one of the following\nphoto\naudio\nvideo\nsticker\nvoice\ndocument\n*/\n\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","x":1060,"y":780,"wires":[["4b3423f9.64fa4c"]]},{"id":"9b50c12.74a3a4","type":"mqtt-broker","name":"192.168.14.204-hik","broker":"192.168.14.204","port":"1883","clientid":"","usetls":false,"compatmode":true,"keepalive":"60","cleansession":true,"birthTopic":"","birthQos":"0","birthPayload":"","closeTopic":"","closePayload":"","willTopic":"","willQos":"0","willPayload":""}]
```

</p>

**Comment by jakowenko** [comment]
Closing since this is now included in the `1.0.0` release.
